### ISSUE TITLE
BREAKING - Create build command and drop --noDeploy option

### ISSUE THREAD

**Issue by flomotlik** [serverless/serverless#2660]
# This is a Feature Proposal

## Description

Some use cases, especially with CI/CD demand the packages to be used for a deployment to be created beforehand. After #2659 is merged we can create a separate build command with its own lifecycle that can then be included by the deployment command.

`serverless build` without any options will package all artifacts for deployment to Lambda and put them in the `.serverless` folder

`serverless build -f hello` will only package the specified function and put it in the `.serverless` folder.

Both package commands will overwrite the artifact configuration of each function to point to the newly created artifacts so other commands like deploy or deploy function can pick them up and release the artifact. This only works if build runs as a dependency of those other commands.

## Future extensions:

``serverless build --output (or -o)` will package all artifacts and put the artifacts into the specified 

Similar or dependent issues:
* #2659
* #2473
* #2661

**Comment by mthenw** [comment]
Dropping `--noDeploy` is a breaking change so I'm not sure if we can release that in 1.*.

**Comment by flomotlik** [comment]
@mthenw FYI I'm only renaming the milestones once it comes up, basically because before the milestone is planned I don't necessarily know if a breaking change will be happening in it. So in this case it will be renamed and will then be at least 2.0

**Comment by johncmckim** [comment]
I think we need to be really clear on the goals and implementation of this issue. 
It will affect how #2659 is implemented.

The main goal is to support pipeline style deployments. As part of pipelines we need to be able to run a series of steps `lint / unit test` -> `package/build` ->  `integration test` -> `deploy`. The build artefact needs to be consistent between testing and deployment.

This will also allow us to create temporary environments comprised of multiple services for testing purposes. i.e. I need to run integration testing on `service-a` which relies on `service-b`. To do this, I deploy `service-a` and a pre-packaged version of `service-b` to a test environment before running my tests.

It seems to me that there are two distinct options for the implementation of `build`.

## Option 1 - code + serverless.yml
This technique will require the user to modify `serverless.yml`. The user will need to use the `package.artifact` setting plus options to conditionally set it. For example

```
package:
  artifact: ${opt:userArtifactOption}
```

The first step in the pipeline would execute `build`. The command will zip the code and `serverless.yml` into a package. This could then be stored as an artefact in a build system.

The second step in pipeline would extract the code + `serverless.yml`  and use `sls deploy --userArtifactOption /path/to/code.zip`. This would create the CF templates on deploy and deploy them to  AWS.

**Advantages**
* Uses existing features so it might be easier to implement
* More configuration options - (this is also a disadvantage)

**Disadvantages**
* Requires user config in `serverless.yml`
* CF templates are built on every deploy
* CF templates could change between deployments - a bug or version change could cause this
* The package is not independently deployable i.e. you couldn't just upload the CF + code to AWS yourself. (This isn't much of a disadvantage though)

## Option 2 - code + CF templates
The first step in the pipeline will execute `build`. The command will zip the contents of `.serverless`. This will include code + create & update CF stacks.

The second step in the pipeline would retrieve the artefact and use `sls deploy --artifact /path/to/artifact.zip`. This would skip the package step, extract the artifacts to `.serverless` and deploy the existing code + CF templates.

**Advantages**
* Better user experience - no config in `serverless.yml`
* CF templates are built only once
* CF templates won't change between deployments
* The package not independently deployable 

**Disadvantages**
* `serverless.yml` isn't available to `sls` which will will require some core changes
* Support for deploying existing artifacts will have to be added to `deploy`
* Plugins that deploy outside of CF wouldn't work i.e. [external sns](https://github.com/silvermine/serverless-plugin-external-sns-events). I personally think they should use custom CF resources and not act outside of CF.

## Summary
Clearly I'm in favour of option 2. I think it's a better user experience and more consistent. But it's probably a harder change to make. Happy to hear different thoughts on this.

**Comment by laardee** [comment]
@johncmckim, for me, the option 2 would be much better. I'm currently working on a project where I deploy zip + CloudFormation files (using `--noDeploy`) with Jenkins. Those files are then used later when the whole stack is deployed to AWS with Ansible.

**Comment by rowanu** [comment]
Not sure if can go with either of your options @johncmckim; Option 2 sounds closer to what I imagined, but I'm confused about your proposal details:
* Why does deploy require an `--artifact` argument? Wouldn't it just deploy the (pre-generated) artifacts in `.serverless/`?
* Why would `serverless.yml` not be available to `sls`? I don't think `sls` should be doing anything without a `serverless.yml`, in general...

I definitely don't think this change should _require_ changes/additions to `serverless.yml` - it should be as transparent as possible to users who don't care about the build step i.e. they can continue to build and deploy in a single command if they want (which is something I often want to do). Only users who care about the `build` step need to know/worry about it. There may be advanced options, but there should be enough sensible defaults that no changes are required.

In my mind I had imagined _additional_ commands to facilitate this:

```bash
sls build ... # builds template and function packages in .serverless/
sls deploy ... # builds AND deploys the service, just like now
sls push ... # ONLY deploys the service, based on existing artifacts in .serverless/ (and fails if none present)
```

To be clear, I'm not in love with the name `push` but I'm just putting it out there to differentiate it from the other commands. This would be backwards compatible - you could continue to support `deploy --noDeploy` to avoid a version bump.

**Comment by johncmckim** [comment]
@rowanu that's a pretty interesting solution.

I was including an `--artifact` argument so deploy could just take a package. It would be responsible for extracting it and putting it into `.serverless` if that's what was required.

I think you're right about `.serverless.yml` having to be there. My thinking behind that was because the built package needs to remain unmodified by deploy. This means the `package`, `function` or `resources` config in `serverless.yml` are not used when deploying a built package. Though upon reflection, it might still need to check the `provider` settings.

I like the UX of `build` & `push` (or another name) and having `deploy` coordinate those two commands. If #2659 is implemented `deploy` would just have `build` as a before command and `push` as an after command.

**Comment by nicka** [comment]
+1 for the 'build' command! However not a big fan of the 'push' command, sounds a bit like only uploading your build.

**Comment by rowanu** [comment]
Yeah @nicka, I was never happy with that name I just wanted something that was different.

After seeing discussion on [the forum](http://forum.serverless.com/t/deploy-api-gateway-individually/824) I think there might even be scope for another command to just update configuration (but not function code):

```bash
sls build ... # builds template and function packages in .serverless/
sls deploy ... # builds AND deploys the service, just like now
sls push ... # ONLY deploys the service, based on existing artifacts in .serverless/ (i.e. uploads zips, etc)
sls update ... # ONLY updates the service CFN, but doesn't upload function code
```

The `update` command would benefit services with large/many functions, to save you uploading them when you've just made a configuration change.

**Comment by johncmckim** [comment]
@rowanu I certainly think a command to just update CFN would be nice. But I think it's a separate issue to this. I see that as being a helpful development command but not useful for a CI/CD pipeline.

@flomotlik @ac360 @pmuens this issue is `priority/P1` but keeps getting pushed back. I personally think that it's really important issue. CI/CD is often one of the first questions I get from people getting into serverless. It's easy to run simple CI + deployments. But, it's not easy to build a stage based CI/CD pipeline right now. A stage base CI/CD pipeline is important for large 

I imagine no work is happening because there's no decision on what will actually be implemented yet. It would be great if we could discuss this issue further and try come to an agreement of what will be implemented.

If we can do that then someone will be able to contribute a PR. I'm happy to contribute a PR after the holidays.

What are your thoughts on the options discussed here?

**Comment by nikgraf** [comment]
hey @johncmckim, last week I started to work on proposal to tackle multiple issues at once and still have a consistent and sane experience. I presented it internally and some more ideas were coming in. I wanna share it asap, but needs some refinement first. Definitely going to share this here!

The issues we try to solve:
- programmatically usage of the framework e.g. a json respons from info is much needed, but basically from any command
- splitting up deploy into multiple steps (what I noticed here is that the AWS CLI `aws package` command packages, uploads and creates a new SAM file. Still need to investigate why they made this decision - input here is very welcome!)
- find a design that works well for multiple cloud providers

**Comment by pmuens** [comment]
@nikgraf thanks for the update!

> splitting up deploy into multiple steps (what I noticed here is that the AWS CLI aws package command packages, uploads and creates a new SAM file. Still need to investigate why they made this decision - input here is very welcome!)

I thought about that as well and my assumption is that they upload it because they need to have an S3 bucket in place. So they create one in order to replace the `CodeUri` property in the template file and upload it right away (because they have the bucket anyway). Just an assumption but I guess that it has something todo with the bucket for deployment...

**Comment by johncmckim** [comment]
@nikgraf looking forward to hearing it.

@pmuens not sure why they upload it to S3. I'd prefer that didn't happen. We should be able to store artefacts in a place that suits our build pipelines. That might be S3, or it might be in Octopus Deploy, Team City, Jenkins ect. I think uploading to S3 makes more sense as part of the deploy stop rather than the package step.

**Comment by jhovell** [comment]
@johncmckim just curious what the advantages of putting Cloudformation templates in TeamCity, etc? It is a requirement of AWS Lambda to host code packages in S3, and since Cloudformation templates have to refer to refer to code packages by bucket name it seems hard to host artifacts outside of S3 unless they contain placeholders or can only be deployed to a predefined bucket name. If it were possible to deploy a Lambda from somewhere other than S3 or use Cloudformation for something beside AWS resources I could see how this makes sense, but within the scope of Cloudformation/Lambda it seems to me like extra moving parts & complexity to store artifacts outside of S3. 
1. It brings up questions about needing to publish artifacts to S3 each time (or determine if they are already there) before deploying.
2. It is more difficult to answer the question did the exact same artifact that was just deployed to staging just get deployed to prd, etc

This could all be overcome with additional tooling but just curious why it would be valuable to track these artifacts in a 3rd party artifact manager

**Comment by vibhakant** [comment]
@jhovell I agree with you that there is this additional step of publishing artifacts to S3 each time, however if there is a 3rd party artifact manager then the need to share the same S3 bucket in all the environments goes away. Each environment deployment will fetch the artifacts from the artifact manager and then performs the steps needed to deploy that artifact which happens to be 'Upload to S3 bucket' + other steps in this case. As for tracking if the same artifact just got deployed, I believe that is no different than making sure artifact location is correctly communicated while promoting the build whether its an S3 location or a 3rd party artifact location. Do you think differently?

Context: John and I are colleagues and have been talking about using S3 bucket itself as the artifact manager and sharing that bucket across all environments. Since CF templates have to refer to code packages stored in the bucket this takes an additional step away and technically becomes part of building and archiving artifacts.

**Comment by johncmckim** [comment]
@jhovell @vibhakant it's not that TeamCity ect is better than S3. My point is that unless serverless wants to become an artefact storage too (which is shouldn't), it needs to be storage independent. Serverless should just package and deploy artefacts not manage the artefact storage.

@jhovell you do make a good point about the packages + templates. Based on your points, I think `sls package` should upload code to S3 and include the references in the CloudFormation template. The CloudFormation template is then the only artefact required to deploy the service. Is that what you were thinking @pmuens @nikgraf ? 

This is the kind of CI/CD pipeline I want to support. (You might need to click on it to view it propertly). My point is that you should be able to build this with whatever CI/CD pipeline tool you currently use. You should also be able to use that tool's artefact storage and not be tied to the provider through the serverless framework.

![ci deploy](https://cloud.githubusercontent.com/assets/1297227/21408562/68d7aaca-c821-11e6-9769-8dbba779c4f1.png)

To be extra clear, I don't want to remove the ability to deploy with a single command. (People keep asking this when I mention this issue to them). Just support two extra scenarios.

As @rowanu said, something like this
```
sls package ... # packages code & uploads to S3. Creates CF template. Both stored in .serverless/
sls deploy ... # builds AND deploys the service, just like now
sls push or sls deploy --artefact xxx ... # ONLY deploys the service, uploads the CF template to S3 and starts an update
```

**Comment by vibhakant** [comment]
@johncmckim 

Just to be clear the CloudFormation template created will be parameterized right? At the moment there is no way for me to parameterize the template for re-use in all environments. For example domain name for a basepathmapping of an API can be different based on the environment, but at the moment I need to know the environment to deploy 'before' I create the CF template. I am guessing that would be changed and sls deploy --artefact x or sls push will take in the --parameters-override like CF deploy does?

**Comment by rowanu** [comment]
> To be extra clear, I don't want to remove the ability to deploy with a single command. (People keep asking this when I mention this issue to them). Just support two extra scenarios.

Yeah, if that wasn't clear before I also don't want to change/remove the **existing** `deploy` command - I want additional functionality/commands.

**Comment by laardee** [comment]
@johncmckim, you've described exactly the flow that I'm using with Jenkins, Ansible and Artifactory. :smile: As @vibhakant states, the parametrization is also needed. Currently I'm using custom post deploy plugin that just replaces regions and stages from cloudformation json with ansible variable syntax. Which is not very sleek approach.

**Comment by arabold** [comment]
@rowanu brings up an important point with the CF templates: For the CI system to work properly the package needs to be independent from the target environment to deploy to. Otherwise a package can never be promoted from dev to prod but would have to be re-packaged.

It might seem a bit off topic here, but right now I'm worried about the way sls deploys functions and environments in general: Every stage is using a separate CloudFormation script that typically has no overlaps with other stages. While this seems a good idea at first it also brings a couple of problems: For example APIG _does_ support stages but the current setup of Serverless wouldn't allow us to make us of it, as the APIG project is owned by the stage itself. We *have to* have two APIG projects because we have two CF templates.

To come back on topic with the `build` and `deploy` command: In an ideal world I would want to split deployment into *three steps*:
1. Setup the project itself. This is a one-time step and prepares resources shared between stages that don't have to be recreated or updated on every deploy. For example this could be the S3 bucket we use for artifacts (no reason to have a bucket for every stage IMO), the API Gateway project and even the Lambda function itself (as every deploy would only add a new version, stages could be aliases #2411 ).
2. Package a release independent from the target environment.
3. Deploy/push the package to the target environment.

**Comment by jhovell** [comment]
@arabold I'd be a bit leery of needing to rely on APIG stages. A lot of people deploy production in a separate AWS account as there are limits to the type of authZ you can do with IAM.

I like the idea of a 3 stage approach at least for the artifact bucket (this can be managed across AWS accounts fine with IAM) but I would think it would be important to not rely on APIG stages (what do they allow for anyway other than just sugar in AWS console UI?).

**Comment by rowanu** [comment]
Some great points there @arabold. I totally agree about package and environment needing to be different - good to call it out.

While I can understand different APW GW stages sharing the same bucket (because they share the same API GW Project), I currently think it's a broken model - I'm hoping to be proved wrong, but so far stages have been nothing but trouble for me.

I would go so far as to suggest that we might need to avoid API GW Stages! API GW is the only service that supports them, and it doesn't translate well in to the other services.

Different "environments" (as opposed to stages) *must* have different supporting infrastructure (e.g. buckets, etc). No large organisation/enterprise will consider deploying non-production and production environments in the same AWS account or with shared resources; I would be concerned if a business of any size were doing it. Using different accounts to limit "blast radius" is simply the recommended way to do things in AWS, for a variety of reasons.

Unfortunately, right now the idea of stages and environments are mixed, and I think it complicates this (and other) problem spaces.

**Comment by arabold** [comment]
Don't want to drift too far off topic (which is splitting deploy into multiple steps) and I agree that you need to be wary when prod and dev isn't strictly separated. "dev" and "prod" might not be good examples given by me in first place. Instead think of "staging" and "prod", or blue-green deployment where you actually have to share some basic shared resources. Or scenarios in which you want to do split testing between two stages. My reasoning behind having multiple stages with shared Lambdas and APIG is explained more in the comment here: https://github.com/serverless/serverless/issues/2411#issuecomment-267663782.

My point is that Serverless should allow certain flexibility and not tie CloudFormation and stages strictly together. I think we all agree that a deployment package _needs_ to be independent from the target stage it's gonna be deployed to.

**Comment by arabold** [comment]
I was thinking about this again and tried to summarize some of my thoughts below. The term "stage" is used although you could also name it "environment" or "flavor". Don't think of it as "prod" vs. "dev" though! I'm thinking more of blue-green deployment and user-specific stages where every engineer can temporarily use his a dedicated stage for a code isolation without having to replicate a whole underlying infrastructure.

Here's how I can see this work without jeopardizing all sls design principles:

1. Separate CloudFormation into 2+ stacks: 
    * A _project stack_ for the project itself that contains components sharable across all stages. This could include the deployment S3 bucket and even the Lambda function and API Gateway definition. This could be optional but I do believe it makes sense as soon as you're talking about more stages than just dev and prod. A stage would be mapped as an "alias" in Lambda functions and as a "stage" in API Gateway. There are more recently introduced services that are supporting aliases/stages, e.g. Lex. The purpose of the project stack is to allow reusing those services in different stages.
    * Plus _one stack per every stage_. That's the one which is used during deployment of a stage.
2. The YAML config should allow putting resources in either the project stack or individual stage stacks. This gives the developer the flexibility to decide which resources can be shared across stages and which shouldn't.
3. Split `sls deploy` into three commands as described in previous comments.
    * A new `prepare` or the `init` command perform the initial project setup and creates the `project stack` in CloudFormation.
    * The `build` will package the code into a ZIP that's usable regardless of the target stage. This can reference or include information from the `project stack` however.
    * `deploy` takes the ZIP file and deploys it using a the dedicates stage specific CloudFormation stack.

**Comment by arabold** [comment]
It is worthwhile noting that alias support as outlined above has been implemented by @HyperBrain as standalone plugin. This addresses most of my initial concerns and I'd like to encourage everybody on this thread to help beta testing: https://github.com/serverless/serverless/issues/2411#issuecomment-283772297
